### PR TITLE
updated infra to match the requests for the new validator

### PR DESCRIPTION
**Pull request by YuvHayun**
## Related Issues
fixes:

## Description
As a part of this PR:
- Added support for save() method to save changes made to the content_item in the original file.
- Added fields that will become handy for validations such as pack's ignored errors and support level.
- Added BaseContentModel class to differentiate between graph objects that are validatable and those that are not (such as commands).
- Added the functionality to create a BaseContent object from a specific git commit.

### CODE REVIEW

File: demisto_sdk/commands/common/tools.py [hunk 1]
```diff
@@ -834,24 +834,26 @@ def get_file(
...
-
     if raise_on_error = False, this function will return empty dict
     """
     if clear_cache:
         get_file.cache_clear()
     file_path = Path(file_path)  # type: ignore[arg-type]
+    if git_sha:
+        if file_path.is_absolute():
+            file_path = file_path.relative_to(get_content_path())
+        return get_remote_file(str(file_path), tag=git_sha)
```
Comment (ilaner) [inline demisto_sdk/commands/common/tools.py:849]: please provide `return_content` to the function

File: demisto_sdk/commands/common/tools.py [hunk 2]
```diff
@@ -3974,3 +3982,84 @@ def parse_int_or_default(value: Any, default: int) -> int:
...
+                if "[" in key and "]" in key:
+                    # Handle list indexing
+                    list_key, index = key.split("[")
+                    index = int(index.strip("]"))  # type: ignore
+                    temp_obj = temp_obj[list_key][index]
+                else:
+                    temp_obj = temp_obj[key]
+            except (AttributeError, KeyError, IndexError):
+                success = False
+                continue
```
Comment (ilaner) [inline demisto_sdk/commands/common/tools.py:4013]: why should we continue and not return?

File: demisto_sdk/commands/common/tools.py [hunk 3]
```diff
@@ -3974,3 +3982,84 @@ def parse_int_or_default(value: Any, default: int) -> int:
 
 def get_all_repo_pack_ids() -> list:
     return [path.name for path in (Path(get_content_path()) / PACKS_DIR).iterdir()]
+
+
+def get(obj: dict, paths: Union[str, List[str]], defaultParam=None):
```
Comment (ilaner) [inline demisto_sdk/commands/common/tools.py:3987]: let's rename to `get_value`

File: demisto_sdk/commands/content_graph/content_graph_builder.py [hunk 4]
```diff
@@ -70,9 +70,8 @@ def _create_content_dtos(self, packs: Optional[List[str]]) -> List[ContentDTO]:
             leave=True,
         ) as progress_bar:
             for packs_batch in more_itertools.chunked(packs_to_parse, PACKS_PER_BATCH):
-                repository_parser.parse(packs_batch)
+                repository_parser.parse(packs_batch, progress_bar)
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/content_graph_builder.py:73]: revert

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 5]
```diff
@@ -86,7 +87,7 @@ def __new__(
         return model_cls
 
 
-class BaseContent(ABC, BaseModel, metaclass=BaseContentMetaclass):
+class BaseContentModel(ABC, BaseModel, metaclass=BaseContentMetaclass):
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/base_content.py:90]: let's rename:
This base class will be `BaseContent`
The subclass will be `BaseContentWithPath`

@dantavori what do you think?
The idea is to create a class between `BaseContent` and the Pack/Content (without commands), so we can use the `path` attribute

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 6]
```diff
@@ -130,6 +131,14 @@ def __getstate__(self):
             "__fields_set__": self.__fields_set__,
         }
 
+    @abstractmethod
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/base_content.py:134]: this function should be only on the subclass (BaseContentWithPath)

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 7]
```diff
@@ -160,32 +169,84 @@ def to_dict(self) -> Dict[str, Any]:
...
+        if relationship.content_item_to == self:
+            # skip adding circular dependency
+            return
+        self.relationships_data[relationship_type].add(relationship)
+
+
+class BaseContent(BaseContentModel):
+    mapping: dict = Field({}, exclude=True)
+    path: Path
+    git_status: Optional[str]
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/base_content.py:196]: use enum for git status instead of str

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 8]
```diff
@@ -160,32 +169,84 @@ def to_dict(self) -> Dict[str, Any]:
...
+    old_file_path: Optional[Path]
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/base_content.py:197]: just `old_path`

File: demisto_sdk/commands/content_graph/objects/content_item.py [hunk 9]
```diff
@@ -183,6 +200,18 @@ def handler(self) -> XSOAR_Handler:
     def data(self) -> dict:
         return get_file(self.path, keep_order=False)
 
+    @property
+    def original_data(self) -> dict:
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/content_item.py:204]: can we think about a better name?

File: demisto_sdk/commands/content_graph/objects/content_item.py [hunk 10]
```diff
@@ -183,6 +200,18 @@ def handler(self) -> XSOAR_Handler:
...
+        return get_file(self.path, keep_order=True)
+
+    def save(self):
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/content_item.py:207]: there is some code duplication between the `save` methods in Pack and in Content Item, which could lead to bugs in the future. There is even a bug here, since you miss handling the `marketplaces` field.

I suppose implementing a method in the `BaseContentWithPath` which receives the data and the path to write to, and the subclass calls it. For instance:
```
def save(self):
            self._save(self.path, self.original_data)
```
```
def save(self):
      self._save(self.metadata_path, self.original_data)
```
Where we add `metadata_path` and `original_data` to the `Pack` object.

File: demisto_sdk/commands/content_graph/objects/pack.py [hunk 11]
```diff
@@ -497,3 +508,23 @@ def to_nodes(self) -> Nodes:
             self.to_dict(),
             *[content_item.to_dict() for content_item in self.content_items],
         )
+
+    def save(self):
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/pack.py:512]: see the comment in the other `save` impl

File: demisto_sdk/commands/content_graph/parsers/pack.py [hunk 12]
```diff
@@ -304,3 +316,33 @@ def deprecated(self) -> bool:
         ):
             return True
         return False
+
+    def parse_ignored_errors(self):
+        """Sets the pack's ignored_errors field."""
+        self.ignored_errors_dict = get_pack_ignore_content(self.path.name) or {}  # type: ignore
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/parsers/pack.py:322]: why the type ignore here?

File: demisto_sdk/commands/content_graph/parsers/pack.py [hunk 13]
```diff
@@ -238,11 +245,16 @@ def __init__(self, path: Path) -> None:
...
-            self.contributors: List[str] = get_json(path / PACK_CONTRIBUTORS_FILENAME)
+            self.contributors: List[str] = get_json(
+                path / PACK_CONTRIBUTORS_FILENAME, git_sha=git_sha
+            )
         except FileNotFoundError:
             logger.debug(f"No contributors file found in {path}")
         logger.debug(f"Parsing {self.node_id}")
         self.parse_pack_folders()
+        self.parse_ignored_errors()
+        logger.debug(f"Successfully parsed {self.node_id}")
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/parsers/pack.py:256]: delete

File: demisto_sdk/commands/content_graph/parsers/parsing_rule.py [hunk 14]
```diff
@@ -10,13 +12,21 @@
...
+
+    @cached_property
+    def mapping(self):
+        super().mapping.update({"object_id": "id"})
+        return super().mapping
 
     @property
     def object_id(self) -> Optional[str]:
-        return self.yml_data.get("id")
+        return get(self.yml_data, self.mapping.get("object_id", ""))
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/parsers/parsing_rule.py:29]: we can remove all those properties in all the subclasses.
The impl is identical to YAMLContentItemParser.

File: demisto_sdk/commands/content_graph/parsers/repository.py [hunk 15]
```diff
@@ -3,6 +3,8 @@
 from pathlib import Path
 from typing import Iterator, List, Optional
 
+from tqdm import tqdm
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/parsers/repository.py:6]: revert the changes here (handled by another PR)

**Review by ilaner — CHANGES_REQUESTED** [review]
Very good job.
Please see notes

File: demisto_sdk/commands/common/tools.py (hunk 2 above)
Comment (YuvHayun) [reply demisto_sdk/commands/common/tools.py:4013]: Because there're cases where the mapping hold several keys options for one item, so we want to be able to find the one where the value is really at.

File: demisto_sdk/commands/content_graph/parsers/pack.py (hunk 13 above)
Comment (YuvHayun) [reply demisto_sdk/commands/content_graph/parsers/pack.py:256]: you mean only the logger line right?

File: demisto_sdk/commands/content_graph/parsers/pack.py [hunk 16]
```diff
@@ -217,16 +222,18 @@ class PackParser(BaseContentParser, PackMetadataParser):
...
 
-    def __init__(self, path: Path) -> None:
+    def __init__(self, path: Path, git_sha: Optional[str] = None) -> None:
         """Parses a pack and its content items.
 
         Args:
             path (Path): The pack path.
         """
+        if path.name == PACK_METADATA_FILENAME:
+            path = path.parent
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/parsers/pack.py:232]: why is it needed?

File: demisto_sdk/commands/content_graph/parsers/xsiam_dashboard.py [hunk 17]
```diff
@@ -16,28 +15,32 @@ class XSIAMDashboardParser(
...
-            or DEFAULT_CONTENT_ITEM_FROM_VERSION
+    @cached_property
+    def mapping(self):
+        super().mapping.update(
+            {
+                "object_id": [
+                    "dashboards_data[0].global_id",
+                    "dashboards_data[0].id",
+                    "id",
+                    "global_id",
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/parsers/xsiam_dashboard.py:33]: ?

File: demisto_sdk/commands/content_graph/tests/test_data/script.yml
Comment (dantavori) [inline demisto_sdk/commands/content_graph/tests/test_data/script.yml:1]: where is it used?

**Review by dantavori — COMMENTED** [review]
good job, need to update the documentation with this changes, here (under content_graph/README.md and confluence)
please go over the fields/mapping again before you merge to ensure we didn't miss anything

File: demisto_sdk/commands/content_graph/tests/test_data/script.yml
Comment (YuvHayun) [reply demisto_sdk/commands/content_graph/tests/test_data/script.yml:1]: It's a part of the testings for the refactor tests, I can add it on the other PR if you prefer

File: demisto_sdk/commands/content_graph/parsers/xsiam_dashboard.py (hunk 17 above)
Comment (YuvHayun) [reply demisto_sdk/commands/content_graph/parsers/xsiam_dashboard.py:33]: there are multiple options for where this key might be.. get_value will know how to find the right one.

File: demisto_sdk/commands/content_graph/parsers/pack.py (hunk 16 above)
Comment (YuvHayun) [reply demisto_sdk/commands/content_graph/parsers/pack.py:232]: We want to be able to parse a pack & pack_metadata from 'from_path' (-i inputs) so in some cases we receive pack_metadata as an input and this is how we handle this case.

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 18]
```diff
@@ -160,32 +165,102 @@ def to_dict(self) -> Dict[str, Any]:
...
+        raise NotImplementedError()
+
     @staticmethod
     @lru_cache
-    def from_path(path: Path) -> Optional["BaseContent"]:
+    def from_path(
+        path: Path,
+        git_status: Optional[GitStatuses] = None,
+        old_file_path: Optional[Path] = None,
+        git_sha: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (YuvHayun) [reply demisto_sdk/commands/content_graph/objects/base_content.py:240]: We won't call it altogether
But the idea is, that if we rename a file, then we create an object with the path, the git_status=Renamed, old_file_path = the file before the renaming.
Then when we want to create an object for the file before the renaming path = the old_file_path of the other file and git_sha will be the prev_ver (probably master)

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 19]
```diff
@@ -160,32 +165,102 @@ def to_dict(self) -> Dict[str, Any]:
...
+    def add_relationship(
+        self, relationship_type: RelationshipType, relationship: "RelationshipData"
+    ) -> None:
+        if relationship.content_item_to == self:
+            # skip adding circular dependency
+            return
+        self.relationships_data[relationship_type].add(relationship)
+
+
+class BaseContentWithPath(BaseContent):
```
[the remark this answers is not in the document]
Comment (ilaner) [reply demisto_sdk/commands/content_graph/objects/base_content.py:179]: BaseContent: the base class of all the nodes - Packs, Content Items, Commands
BaseContentWithPath: the base class of Packs and Content Items. Because they have `path` they share a lot of behavior

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 20]
```diff
@@ -160,32 +165,109 @@ def to_dict(self) -> Dict[str, Any]:
...
-    def from_path(path: Path) -> Optional["BaseContent"]:
+    def from_path(
+        path: Path,
+        git_status: Optional[GitStatuses] = None,
+        old_file_path: Optional[Path] = None,
+        git_sha: Optional[str] = None,
+    ) -> Optional["BaseContentWithPath"]:
         logger.debug(f"Loading content item from path: {path}")
+        if git_status in (GitStatuses.MODIFIED, GitStatuses.RENAMED):
+            obj = BaseContentWithPath.from_path(path, git_status)
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/base_content.py:244]: please comment what we are doing here, as it is not that obvious

File: demisto_sdk/commands/content_graph/objects/base_content.py [hunk 21]
```diff
@@ -206,39 +288,15 @@ def from_path(path: Path) -> Optional["BaseContent"]:
             logger.error(f"Could not parse content item from path: {path}")
             return None
         try:
-            return model.from_orm(content_item_parser)
+            obj = model.from_orm(content_item_parser)
+            obj.git_status = git_status
```
Comment (ilaner) [inline demisto_sdk/commands/content_graph/objects/base_content.py:292]: let's go back to `return model.from_orm(content_item_parser)`, and set the git_status in the first `if` condition

**Review by ilaner — APPROVED** [review]
Great!

File: demisto_sdk/commands/content_graph/parsers/xsiam_dashboard.py (hunk 17 above)
Comment (dantavori) [reply demisto_sdk/commands/content_graph/parsers/xsiam_dashboard.py:33]: can you please share a path to an existing XSIAM dashboard that has "id" or "global_id" fields in the root level of the json?

Comment (YuvHayun) [reply demisto_sdk/commands/content_graph/parsers/xsiam_dashboard.py:33]: Oh my bad, looking into this again, I got confused because in the super class we'd extract dashboards_data and then call id / global_id.. removing.
